lib/services/application_service.ts

```diff
@@ -43,6 +43,8 @@ export class ApplicationService extends PureService {
         this.onAppIncrementalSync();
       } else if (event === ApplicationEvent.KeyStatusChanged) {
         this.onAppKeyChange();
+      } else if (event === ApplicationEvent.ExpiredAccessToken) {
+        this.onAppExpiredToken();
```

Comment:
Not sure I'm following why this is in application_service? This is an abstract class that clients can subclass from. But it's a client related wrapper, not snjs related wrapper. And there can exist many subclasses of this class, so in this case, `onAppExpiredToken` can be called ten times.

Reply:
Where should I check for when the `SyncEvent.ExpiredAccessToken` event is triggered? Can I do so from within the `api_service` or `session_manager`?

Reply:
I feel like session renewal should be handed completely by the api_service. If a session is expired, it freezes all sync requests within the api_service (or adds them to a resolve queue). Once the session renews, any pending sync requests are executed serially or concurrently, and their original promise resolves. I don't think we need to get the `application` or `sync_service` involved in session management. Take a look at how we preserve the original promises of functions here: https://github.com/standardnotes/snjs/blob/004/lib/services/model_manager.ts#L127

Reply:
The thing is that both `sync_service` and `session_manager` are not available from `api_service`. I need `sync_service` for toggling lock syncing, and I also need `session_manager` to update the session object with the new tokens.

Reply:
The api_service can definitely lock syncing as well if it wanted to.

Reply:
In this case it seems like the best way is to have api_service and session_manager collaborate, but no one else.

Reply:
> The api_service can definitely lock syncing as well if it wanted to.

Should I lock syncing by checking if the session is being renewed? If so, any syncing requests should be queued and processed later.

Reply:
Does it make sense to use `SyncEvent` within `api_service`? I'm thinking of using `this.notifyEvent(SyncEvent.ExpiredAccessToken)` from `http_service` (I will call this method whenever I get the custom HTTP code that is returned when the access token is expired) and in `api_service`, I could have an observer that will trigger my `refreshSession` functionality whenever the `SyncEvent.ExpiredAccessToken` is notified.

Reply:
Nope, http_service is totally agnostic to the functionings of the app. Should definitely not trigger any events. It is stateless. Only sync_service should trigger sync events.

Reply:
Thanks for pointing me into the right direction. I've pushed new changes :+1: